src/parser/jobs/rpr/modules/ImmortalSacrifice.tsx

```diff
@@ -0,0 +1,154 @@
+import BrokenLog from '../../../core/modules/BrokenLog'
+import Checklist, {Requirement, Rule} from '../../../core/modules/Checklist'
+import {Data} from '../../../core/modules/Data'
+import {DISPLAY_ORDER} from './DISPLAY_ORDER'
+
+interface ImmortalSacrificeWindow {
+	startTime: number,
+	endTime?: number,
+	/**
+	 * If this window was closed, by consuming the stacks, or it expired
+	 */
+	closed: boolean,
+	stackCount: number,
+	consumed: boolean,
+}
+
+export default class ImmortalSacrifice extends Analyser {
+
+	static override handle = 'immortalsacrifice'
+	static override title = t('rpr.immortal-sacrifice.title')`Immortal Sacrifice stacks consumed`
+	static override debug = false
+
+	@dependency private brokenLog!: BrokenLog
+	@dependency private data!: Data
+	@dependency private checklist!: Checklist
+
+	private windows: ImmortalSacrificeWindow[] = []
+
+	override initialise() {
+		super.initialise()
+
+		const playerFilter = filter<Event>().source(this.parser.actor.id)
+
+		this.addEventHook(playerFilter.type('statusApply').status(this.data.statuses.IMMORTAL_SACRIFICE.id), this.onImmortalSacrificeStack)
+		this.addEventHook(playerFilter.type('statusRemove').status(this.data.statuses.IMMORTAL_SACRIFICE.id), this.onImmortalSacrificeExpiration)
+		this.addEventHook(playerFilter.type('action').action(this.data.actions.PLENTIFUL_HARVEST.id), this.onPlentifulHarvest)
+
+		this.addEventHook('complete', this.onComplete)
+	}
+
+	private onImmortalSacrificeStack(event: Events['statusApply']) {
+		let window = this.getOpenWindow()
+
+		if (window === null) {
+			window = {
+				startTime: event.timestamp,
+				closed: false,
+				consumed: false,
+				stackCount: 0,
+			}
+
+			this.windows.push(window)
+		}
+
+		window.stackCount = event.data ?? window.stackCount + 1
+	}
+
+	private onPlentifulHarvest(event: Events['action']) {
+		const window = this.getOpenWindow()
+
+		if (window === null) {
+			this.brokenLog.trigger(this, 'PH without immortal sacrifice stack',
+				<Trans id="rpr.immortal-sacrifice.trigger.no-immortal-sacrifice">
+					<DataLink action="PLENTIFUL_HARVEST"/> was executed but there were
+					no <DataLink status="IMMORTAL_SACRIFICE"/> stacks.
+				</Trans>)
+			return
+		}
+
+		this.debug('Closing window because plentiful harvest were used', window)
+
+		window.closed = true
+		window.endTime = event.timestamp
+		window.consumed = true
+	}
+
+	private onImmortalSacrificeExpiration(event: Events['statusRemove']) {
+		const window = this.getOpenWindow()
+
+		// valid case, means it was consumed
+		if (window === null) {
+			return
+		}
+
+		this.debug('Closing window that was expired before consumption', window)
+
+		window.closed = true
+		window.consumed = false
+		window.endTime = event.timestamp
+	}
+
+	private getOpenWindow() {
+		if (this.windows.length > 0) {
+			const w = this.windows[this.windows.length - 1]
+
+			if (!w.closed) {
+				return w
+			}
+
+		}
+
+		return null
+	}
```

Comment:
This should have a proper return type, but realistically this pattern is awkward and error prone. You might wanna try having a class var for the current window and just have a stopAndSave function that closes and pushes it to a history var.

Reply:
Reworked